File: diversity_sampler_r/metrics.R
# diversity_sampler_r/metrics.R
# R equivalent of metrics.py

diversity_metrics <- function(X, idxs) {
  idxs <- as.integer(idxs)
  if (length(idxs) == 0) {
    return(list(k = 0))
  }
  
  Xs <- X[idxs, , drop = FALSE]
  
  # mean pairwise cosine similarity
  Xn <- Xs / pmax(sqrt(rowSums(Xs^2)), 1e-12)
  S <- Xn %*% t(Xn)
  n <- length(idxs)
  
  # Get upper triangle indices (excluding diagonal)
  iu <- upper.tri(S, diag = FALSE)
  mean_pairwise_cos <- if (sum(iu) > 0) mean(S[iu]) else 1.0
  
  # nearest-neighbor cosine distance
  S_diag <- S
  diag(S_diag) <- -Inf
  nn_sim <- apply(S_diag, 1, max)
  mean_nn_cos_dist <- mean(1.0 - nn_sim)
  
  return(list(
    k = n,
    mean_pairwise_cosine = mean_pairwise_cos,
    mean_nn_cosine_distance = mean_nn_cos_dist
  ))
} 

File: text_diversity_demo/metrics.py
from __future__ import annotations
import numpy as np

def diversity_metrics(X: np.ndarray, idxs):
    idxs = list(idxs)
    if len(idxs) == 0:
        return {"k": 0}
    Xs = X[idxs]
    Xn = Xs / np.maximum(np.linalg.norm(Xs, axis=1, keepdims=True), 1e-12)
    S = Xn @ Xn.T
    n = len(idxs)
    iu = np.triu_indices(n, k=1)
    mean_pairwise_cos = float(S[iu].mean()) if iu[0].size>0 else 1.0
    np.fill_diagonal(S, -np.inf)
    nn_sim = S.max(axis=1)
    mean_nn_cos_dist = float((1.0 - nn_sim).mean())
    return {
        "k": n,
        "mean_pairwise_cosine": mean_pairwise_cos,
        "mean_nn_cosine_distance": mean_nn_cos_dist,
    }
